deeplearning4j/deeplearning4j-data/deeplearning4j-datasets/src/main/java/org/deeplearning4j/datasets/fetchers/MnistDataFetcher.java

```diff
@@ -232,8 +236,19 @@ public void fetch(int numExamples) {
     public void reset() {
         cursor = 0;
         curr = null;
-        if (shuffle)
-            MathUtils.shuffleArray(order, rng);
+        if (shuffle) {
+            if((train && numExamples < NUM_EXAMPLES) || (!train && numExamples < NUM_EXAMPLES_TEST)){
+                //Shuffle only first N elements
+                if(firstShuffle){
+                    MathUtils.shuffleArray(order, rng);
```

Comment:
Maybe just do this in the last line of the constructor rather than call `reset()`. Would save some code complexity and avoid problems if a client attempts to override the public `reset()` method.